Implement trusted types integration with script elements IDL

lukewarlow opened the pull request:
#### 4e8f9225aef76436cdf18015c6037d5810d87fae
<pre>
Implement trusted types integration with script elements IDL
<a href="https://bugs.webkit.org/show_bug.cgi?id=269718">https://bugs.webkit.org/show_bug.cgi?id=269718</a>

Reviewed by Ryosuke Niwa and Darin Adler.

Implement shadow properties for textContent and innerText on HTMLScriptElement.
Implement enforcement on src and text as well as the shadowed properties.

This doesn&apos;t implement any changes to the prepareScript function that will be addressed in a follow up.

Spec: <a href="https://w3c.github.io/trusted-types/dist/spec/#enforcement-in-scripts">https://w3c.github.io/trusted-types/dist/spec/#enforcement-in-scripts</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::trustedTypeCompliantString):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::setText):
(WebCore::HTMLScriptElement::setTextContent):
(WebCore::HTMLScriptElement::setInnerText):
(WebCore::HTMLScriptElement::src const):
(WebCore::HTMLScriptElement::setSrc):
(WebCore::HTMLScriptElement::addSubresourceAttributeURLs const):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.idl:
* Source/WebKitLegacy/mac/DOM/DOMHTMLScriptElement.mm:
(-[DOMHTMLScriptElement setText:]):

Canonical link: <a href="https://commits.webkit.org/278343@main">https://commits.webkit.org/278343@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1791df037c2dcc0108f9cd7ecb870f7b481fc8d1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50039 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53282 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/716 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35726 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/285 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40826 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52137 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27174 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43041 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21942 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24598 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/282 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8409 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46584 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/329 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54864 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25133 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/454 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48226 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26391 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43242 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47384 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11024 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27427 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26120 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1788,6 +1788,27 @@ void Node::setTextContent(String&& text)
+    return textContent();
+}
+
+ExceptionOr<void> Node::setTextContentForBinding(std::optional<std::variant<String, RefPtr<TrustedScript>>> value)
```
This is not custom binding code. We just have a version defined for JS API.

lukewarlow replied:
Is the forBinding name still correct for the method?

rniwa replied:
Yes.

rniwa commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -85,6 +88,9 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
     Ref<Element> cloneElementWithoutAttributesAndChildren(Document&) final;
+
+    // https://w3c.github.io/trusted-types/dist/spec/#dom-htmlscriptelement-scripttext-slot
+    String m_scriptText;
```
Why is this needed?

lukewarlow replied:
A follow up PR will make use of this slot for some trusted types enforcement. Happy to remove the slot from this PR though.

rniwa replied:
Please remove that from this PR.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -200,6 +201,8 @@ class Node : public EventTarget {
     WEBCORE_EXPORT const AtomString& lookupNamespaceURI(const AtomString& prefix) const;
 
+    WEBCORE_EXPORT std::optional<std::variant<String, RefPtr<TrustedScript>>> textContentForBinding();
+    WEBCORE_EXPORT virtual ExceptionOr<void> setTextContentForBinding(const std::optional<std::variant<String, RefPtr<TrustedScript>>>&);
```
Can we use three overloads here instead of `std::variant`? Seems like it would be higher performance and easier to read the code. I guess this depends on what the bindings generator makes.

lukewarlow replied:
I don't think that works because the IDL definition uses a union which translates to a variant. I don't think WebKit's IDL generator supports the concept of overloading?

lukewarlow replied:
I've looked into this and I can't find a way to use overloading for attributes, if these were functions it'd be easy but because they're attributes it seems you can only really have one getter and setter.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -66,6 +67,8 @@ class HTMLElement : public StyledElement {
     WEBCORE_EXPORT String title() const final;
 
+    WEBCORE_EXPORT std::variant<String, RefPtr<TrustedScript>> innerTextForBinding();
+    WEBCORE_EXPORT virtual ExceptionOr<void> setInnerTextForBinding(const std::variant<String, RefPtr<TrustedScript>>&);
```
Can we use two overloads here instead of `std::variant`? Seems like it would be higher performance and easier to read the code.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -66,6 +67,8 @@ class HTMLElement : public StyledElement {
 
     WEBCORE_EXPORT String title() const final;
 
+    WEBCORE_EXPORT std::variant<String, RefPtr<TrustedScript>> innerTextForBinding();
```
Can this just return `String`? Can we make this an inline in the header file?

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,6 +114,45 @@ String HTMLScriptElement::crossOrigin() const
+
+    auto stringValue = stringValueHolder.releaseReturnValue();
+
+    return setInnerText(WTFMove(stringValue));
```
Should just do this in a single line and won’t need `WTFMove`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,6 +114,45 @@ String HTMLScriptElement::crossOrigin() const
+
+    auto stringValue = stringValueHolder.releaseReturnValue();
+
+    setTextContent(WTFMove(stringValue));
```
Should just do this in a single line and won’t need `WTFMove`.

lukewarlow replied:
Ah yeah, this had an intermediate line originally but I'll change it to WTFMove in the follow up where it makes sense.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,6 +114,45 @@ String HTMLScriptElement::crossOrigin() const
+{
+    ExceptionOr<String> stringValueHolder = WTF::switchOn(
+        value.value_or(emptyString()),
+        [this](const String& str) -> ExceptionOr<String> {
```
WebKit coding style uses words, not abbreviations like `str`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,6 +114,45 @@ String HTMLScriptElement::crossOrigin() const
+{
+    ExceptionOr<String> stringValueHolder = WTF::switchOn(
+        value,
+        [this](const String& str) -> ExceptionOr<String> {
```
WebKit coding style uses words, not abbreviations like `str`.

darinadler replied:
This was marked as resolved, but was not resolved. Please don’t use `str` in the new code.

lukewarlow replied:
Apologies for the oversight will make sure to fix this.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -539,6 +539,26 @@ void HTMLElement::setDir(const AtomString& value)
+{
+    String stringValue = WTF::switchOn(
+        value,
+        [](const String& str) -> String {
```
WebKit coding style uses words, not abbreviations like `str`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -46,6 +46,9 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
     WEBCORE_EXPORT String crossOrigin() const;
 
+    WEBCORE_EXPORT ExceptionOr<void> setInnerTextForBinding(const std::variant<String, RefPtr<TrustedScript>>&) override;
+    WEBCORE_EXPORT ExceptionOr<void> setTextContentForBinding(const std::optional<std::variant<String, RefPtr<TrustedScript>>>&) override;
```
These should be `final` instead of `override`.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -200,6 +201,8 @@ class Node : public EventTarget {
     WEBCORE_EXPORT const AtomString& lookupPrefix(const AtomString& namespaceURI) const;
     WEBCORE_EXPORT const AtomString& lookupNamespaceURI(const AtomString& prefix) const;
 
+    WEBCORE_EXPORT std::optional<std::variant<String, RefPtr<TrustedScript>>> textContentForBinding();
```
Can this return `String` or `std::optional<String>` instead of `std::optional<std::variant>`? Depends on what the bindings generator generates for this.

lukewarlow replied:
Based on initial testing locally it seems it can just be a String return didn't think that would work given the union type. I guess it makes sense that you can do it, String is ultimately just a more specific subset of the union type. Will inline it in the header too if it works.

lukewarlow replied:
Turns out this had to be std::optional rather than just String, so I won't inline in header? But at least it's not a variant type anymore.

darinadler replied:
I would still suggest putting it in the header and inlined if you can do that without adding includes to the header.

lukewarlow replied:
Inlined with or without the inline keyword? Just so I know. I'll let the current push ride the pipelines to check there's nothing else to fix before I push this up.

darinadler replied:
The `inline` keyword is needed if the function body is outside the class definition. Otherwise it is not needed.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -31,6 +31,7 @@
 #include "StyleValidity.h"
 #include "TaskSource.h"
 #include "TreeScope.h"
+#include "TrustedScript.h"
```
We should try to avoid adding this header file; it will slow down the build to add another header to a commonly included header like `Node.h`. Instead we should see if we can use a forward declaration.

lukewarlow replied:
A normal forward declaration doesn't work because it's used in a RefPtr.

darinadler replied:
That sounds oversimplified. We can compile `RefPtr<T>` with a forward declaration of `T`, but not necessarily _uses_ of `RefPtr<T>`. So we need to make sure the right `.cpp` files have the include but not include in the `.h` file. We have gotten this wrong before and slowed down the build a lot and then came back and fixed it.

lukewarlow replied:
Possibly it's the combination of RefPtr and variant? It's definitely something I've tried with previous PRs and it not compile

lukewarlow replied:
Okay this specifically does seem to work as a forward declaration, apologies I did try before but must have been doing something wrong or hitting a specific condition. I'll have a look elsewhere and do a follow up if needed for other places.

darinadler reviewed (changes requested):
We need to do performance testing on this change. Some of these may be hot functions in webpages and things like Speedometer.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -200,6 +201,8 @@ class Node : public EventTarget {
     WEBCORE_EXPORT const AtomString& lookupPrefix(const AtomString& namespaceURI) const;
     WEBCORE_EXPORT const AtomString& lookupNamespaceURI(const AtomString& prefix) const;
 
+    inline String textContentForBinding() const { return textContent(); }
```
Don’t need the inline keyword here.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -66,6 +67,8 @@ class HTMLElement : public StyledElement {
 
     WEBCORE_EXPORT String title() const final;
 
+    inline String innerTextForBinding() { return innerText(); }
```
Don’t need the inline keyword here.

lukewarlow commented on Source/WebCore/html/HTMLScriptElement.idl:
```diff
@@ -21,6 +21,8 @@
 ] interface HTMLScriptElement : HTMLElement {
     [CEReactions=NotNeeded] attribute ScriptString text;
+    [CEReactions=NotNeeded] attribute [LegacyNullToEmptyString] ScriptString innerText;
+    [CEReactions=NotNeeded] attribute ScriptString? textContent;
```
Need to investigate why this isn't behaving quite as expected with null values.

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments.html:
```diff
@@ -31,7 +31,7 @@
       [ "abc", "TrustedHTML", "Element innerHTML",
           _ => div.innerHTML = "abc" ],
-      [ "2+2", "TrustedScript", "Node textContent",
+      [ "2+2", "TrustedScript", "HTMLScriptElement textContent",
```
https://github.com/web-platform-tests/wpt/pull/45493

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -25,9 +25,12 @@
 
 #include "HTMLElement.h"
 #include "ScriptElement.h"
+#include "TrustedScript.h"
```
I’d think we could compile `RefPtr<TrustedScript>` without including `TrustedScript.h`. We should be able to forward declare `TrustedScript`, and then include `TrustedScript.h` in the files that call these functions. An exception might be the bindings, which may need to compile calls to these functions without having an easy way to make it include `TrustedScript.h`. We want to work hard, though, to avoid adding more includes to headers because of the effect on build time.

lukewarlow replied:
I believe the usage in the WEBCORE_EXPORT annotated method is what leads to this being required as an import rather than forward declaration. I can get the build error if that helps track down a potential solution?

darinadler replied:
What’s the `setText` caller outside the bindings that requires `WEBCORE_EXPORT`? Maybe we can add an overload that is exported rather than exporting the version that takes a `std::variant`.

lukewarlow replied:
It's the below code from inside WebKitLegacy DOMHTMLScriptElement file. You're probably right though we can just export a different function (any suggestions for naming?) which just takes a string. Trusted types enforcement isn't fully implemented inside of WebKit Legacy anyway so this won't be the only instance of it.

```
- (void)setText:(NSString *)newText
{
    WebCore::JSMainThreadNullState state;
    IMPL->setText(newText);
}
```

lukewarlow replied:
...okay scratch that I've just tried again and it's working, so another change must have got it working (potentially it was the missing import in the WebKit legacy).

Just gonna double check it with const &

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -36,9 +39,16 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
 
     String text() const { return scriptContent(); }
-    WEBCORE_EXPORT void setText(String&&);
+    WEBCORE_EXPORT ExceptionOr<void> setText(std::variant<RefPtr<TrustedScript>, String>);
```
Why doesn’t this take a `&&` or at least a `const&`? We may be able to follow an idiom with a bit less reference count churn.

lukewarlow replied:
So far I've locally changed to use const& and that's working.

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -36,9 +39,16 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
+    WEBCORE_EXPORT ExceptionOr<void> setText(std::variant<RefPtr<TrustedScript>, String>);
+
+    String textContentForBindings() const;
+    ExceptionOr<void> setTextContentForBindings(std::optional<std::variant<RefPtr<TrustedScript>, String>>);
```
Why doesn’t this take a `&&` or at least a `const&`? We may be able to follow an idiom with a bit less reference count churn.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,68 @@ void HTMLScriptElement::didFinishInsertingNode()
+ExceptionOr<void> HTMLScriptElement::setText(std::variant<RefPtr<TrustedScript>, String> value)
 {
-    setTextContent(WTFMove(value));
+    ExceptionOr<String> stringValueHolder = WTF::switchOn(
```
This repeated idiom calls for a helper function. We should pass in the `variant` and the `ASCIILiteral` argument to `trustedTypeCompliantString` and share the code.

lukewarlow replied:
Ive made a new trustedScriptCompliantStringFromVariant function and an equivalent one for trusted script URL (I'll make one for trusted html in a follow up and do some refactoring of existing code). Let me know what you think.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,16 +173,34 @@ String HTMLScriptElement::crossOrigin() const
+        [this](const String& str) -> ExceptionOr<String> {
+            return trustedTypeCompliantString(TrustedType::TrustedScriptURL, *scriptExecutionContext(), str, "HTMLScriptElement src"_s);
+        },
+        [](const RefPtr<TrustedScriptURL>& trustedScriptURL) -> ExceptionOr<String> {
```
Could just use a shorter name like `scriptURL`; no need to emphasize **trusted** in this local variable name, and names that are closer to just one word are typically easier to read. I think we could do the same with the local variables named `trustedScript`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -36,9 +39,16 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
+    ExceptionOr<void> setTextContentForBindings(std::optional<std::variant<RefPtr<TrustedScript>, String>>);
+
+    String innerTextForBindings();
+    ExceptionOr<void> setInnerTextForBindings(const std::variant<RefPtr<TrustedScript>, String>&);
```
Why doesn’t this take a `&&`? We may be able to follow an idiom with a bit less reference count churn.

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -36,9 +39,16 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
 
-    URL src() const;
+    String src() const;
+    ExceptionOr<void> setSrc(const std::variant<RefPtr<TrustedScriptURL>, String>&);
```
Why doesn’t this take a `&&`? We may be able to follow an idiom with a bit less reference count churn.

lukewarlow replied:
I've converted all these methods to use const && and WTFMove where appropriate. Let me know if there's anything else I can do to improve this.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,68 @@ void HTMLScriptElement::didFinishInsertingNode()
+ExceptionOr<void> HTMLScriptElement::setTextContentForBindings(std::optional<std::variant<RefPtr<TrustedScript>, String>> value)
+{
+    ExceptionOr<String> stringValueHolder = WTF::switchOn(
+        value.value_or(emptyString()),
```
Does this do the right thing for a null? Does `trustedTypeCompliantString` handle empty strings in the best possible way? Do we really expect the same behavior for a null as for an empty string? Is there test coverage for this?

lukewarlow replied:
The final subtest of https://wpt.live/trusted-types/block-string-assignment-to-HTMLElement-generic.html tests assigning null to script.textContent, converting this null to an empty string for the enforcement code is also what Chrome does.

darinadler commented on Source/WebCore/dom/TrustedScript.h:
```diff
@@ -29,7 +29,7 @@
 namespace WebCore {
 
-class TrustedScript final : public ScriptWrappable, public RefCounted<TrustedScript> {
+class WEBCORE_EXPORT TrustedScript final : public ScriptWrappable, public RefCounted<TrustedScript> {
```
Instead of exporting this, we can add an overload that takes a `const String&` or a `String&&` and have the Objective-C++ code call that overload rather than the one that takes a `std::variant`. As long as the Objective-C++ code is not constructing the `std::variant` then it doesn’t need to include this header.

darinadler commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -173,6 +173,32 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
     return stringValue;
 }
 
+ExceptionOr<String> trustedScriptCompliantStringFromVariant(ScriptExecutionContext& scriptExecutionContext, std::variant<RefPtr<TrustedScript>, String>&& input, const String& sink)
```
This function doesn’t need `FromVariant` in its name, because of how overloading works in C++. Can we just make it an overload for `trustedScriptCompliantString`?

darinadler commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -173,6 +173,32 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
+    );
+}
+
+ExceptionOr<String> trustedScriptURLCompliantStringFromVariant(ScriptExecutionContext& scriptExecutionContext, std::variant<RefPtr<TrustedScriptURL>, String>&& input, const String& sink)
```
This function doesn’t need `URL` or `FromVariant` in its name, because of how overloading works in C++. Can we just make it an overload for `trustedScriptCompliantString`?

darinadler commented on Source/WebCore/dom/TrustedType.h:
```diff
@@ -53,6 +53,10 @@ WEBCORE_EXPORT ExceptionOr<String> trustedTypeCompliantString(TrustedType, Scrip
 
+ExceptionOr<String> trustedScriptCompliantStringFromVariant(ScriptExecutionContext&, std::variant<RefPtr<TrustedScript>, String>&&, const String& sink);
+
+ExceptionOr<String> trustedScriptURLCompliantStringFromVariant(ScriptExecutionContext&, std::variant<RefPtr<TrustedScriptURL>, String>&&, const String& sink);
```
Not sure this needs to be in a header. Can these helper functions live in `HTMLScriptElement.cpp` instead?

lukewarlow replied:
These will end up being used in other places so I think they're better off in this file.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,44 @@ void HTMLScriptElement::didFinishInsertingNode()
+ExceptionOr<void> HTMLScriptElement::setText(std::variant<RefPtr<TrustedScript>, String>&& value)
 {
-    setTextContent(WTFMove(value));
+    ExceptionOr<String> stringValueHolder = trustedScriptCompliantStringFromVariant(*scriptExecutionContext(), WTFMove(value), "HTMLScriptElement text"_s);
```
I suggest writing `auto` here instead of `ExceptionOr<String>` or refactoring so there is no local variable.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,44 @@ void HTMLScriptElement::didFinishInsertingNode()
+
+ExceptionOr<void> HTMLScriptElement::setTextContentForBindings(std::optional<std::variant<RefPtr<TrustedScript>, String>>&& value)
+{
+    ExceptionOr<String> stringValueHolder = trustedScriptCompliantStringFromVariant(*scriptExecutionContext(), value ? WTFMove(*value) : emptyString(), "HTMLScriptElement textContent"_s);
```
I suggest writing `auto` here instead of `ExceptionOr<String>` or refactoring so there is no local variable.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,44 @@ void HTMLScriptElement::didFinishInsertingNode()
+        return stringValueHolder.releaseException();
+
+    setTextContent(stringValueHolder.releaseReturnValue());
+    return { };
```
This part of the code could be shared with `setTextContentForBindings` if we add a private member function `ExceptionOr<void> setTextContent(ExceptionOr<String>)`.
```
return setText(trustedScriptCompliantString(*scriptExecutionContext(), WTFMove(value), "HTMLScriptElement text"_s));
```

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,44 @@ void HTMLScriptElement::didFinishInsertingNode()
+        return stringValueHolder.releaseException();
+
+    setTextContent(stringValueHolder.releaseReturnValue());
+    return { };
```
This part of the code could be shared with `setTextContentForBindings` if we add a private member function `ExceptionOr<void> setText(ExceptionOr<String>)`.
```
return setTextContent(trustedScriptCompliantString(*scriptExecutionContext(), value ? WTFMove(*value) : emptyString(), "HTMLScriptElement textContent"_s));
```

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -87,9 +88,44 @@ void HTMLScriptElement::didFinishInsertingNode()
+
+ExceptionOr<void> HTMLScriptElement::setInnerTextForBindings(std::variant<RefPtr<TrustedScript>, String>&& value)
+{
+    ExceptionOr<String> stringValueHolder = trustedScriptCompliantStringFromVariant(*scriptExecutionContext(), WTFMove(value), "HTMLScriptElement innerText"_s);
```
I suggest writing `auto` here instead of `ExceptionOr<String>`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,16 +149,26 @@ String HTMLScriptElement::crossOrigin() const
+
+ExceptionOr<void> HTMLScriptElement::setSrc(std::variant<RefPtr<TrustedScriptURL>, String>&& value)
+{
+    ExceptionOr<String> stringValueHolder = trustedScriptURLCompliantStringFromVariant(*scriptExecutionContext(), WTFMove(value), "HTMLScriptElement src"_s);
```
I suggest writing `auto` here instead of `ExceptionOr<String>`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.idl:
```diff
@@ -20,14 +20,16 @@
 ] interface HTMLScriptElement : HTMLElement {
-    [CEReactions=NotNeeded] attribute ScriptString text;
+    [CEReactions=NotNeeded] attribute (TrustedScript or DOMString) text;
+    [CEReactions=NotNeeded, ImplementedAs=innerTextForBindings] attribute (TrustedScript or [LegacyNullToEmptyString] DOMString) innerText;
```
I don’t think we need `ImplementedAs`, we can just overload `setInnerText`. That will save us from having to write `innerTextForBindings` at all.

darinadler commented on Source/WebCore/html/HTMLScriptElement.idl:
```diff
@@ -20,14 +20,16 @@
-    [CEReactions=NotNeeded] attribute ScriptString text;
+    [CEReactions=NotNeeded] attribute (TrustedScript or DOMString) text;
+    [CEReactions=NotNeeded, ImplementedAs=innerTextForBindings] attribute (TrustedScript or [LegacyNullToEmptyString] DOMString) innerText;
+    [CEReactions=NotNeeded, ImplementedAs=textContentForBindings] attribute (TrustedScript or DOMString)? textContent;
```
I don’t think we need `ImplementedAs`, we can just overload `setTextContentForBindings`. That will save us from having to write `textContentForBindings` at all.

darinadler reviewed (changes requested):
This is a good step in the right direction. We can still do this more cleanly. Also, we need to find out exactly what is causing the Speedometer 3 performance regression and make a version that doesn’t make it slower.

lukewarlow commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -36,9 +39,15 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
 
     String text() const { return scriptContent(); }
-    WEBCORE_EXPORT void setText(String&&);
+    WEBCORE_EXPORT void setTextExported(String&&);
```
Not sure what to call this function?

I couldn't leave it as setText else I got a build error. The ambiguity I think is from converting NSString* to an appropriate type, not sure if there's a better way to avoid the ambiguity?

```
In file included from /Users/luke/Development/BrowserEngines/webkit/WebKitBuild/Debug/DerivedSources/WebKitLegacy/unified-sources/UnifiedSource11-mm.mm:3:
/Users/luke/Development/BrowserEngines/WebKit/Source/WebKitLegacy/mac/DOM/DOMHTMLScriptElement.mm:52:11: error: call to member function 'setText' is
      ambiguous
    IMPL->setText(newText);
    ~~~~~~^~~~~~~
In file included from /Users/luke/Development/BrowserEngines/webkit/WebKitBuild/Debug/DerivedSources/WebKitLegacy/unified-sources/UnifiedSource11-mm.mm:3:
In file included from /Users/luke/Development/BrowserEngines/WebKit/Source/WebKitLegacy/mac/DOM/DOMHTMLScriptElement.mm:32:
/Users/luke/Development/BrowserEngines/webkit/WebKitBuild/Debug/WebCore.framework/PrivateHeaders/HTMLScriptElement.h:42:25: note: candidate function
    WEBCORE_EXPORT void setText(String&&);
                        ^
/Users/luke/Development/BrowserEngines/webkit/WebKitBuild/Debug/WebCore.framework/PrivateHeaders/HTMLScriptElement.h:43:23: note: candidate function
    ExceptionOr<void> setText(std::variant<RefPtr<TrustedScript>, String>&&);
```

darinadler replied:
A fix to the ambiguity is to write `IMPL->setText(String(newText))`.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -86,10 +87,39 @@ void HTMLScriptElement::didFinishInsertingNode()
     ScriptElement::didFinishInsertingNode();
 }
 
+void HTMLScriptElement::setTextExported(String&& value)
```
The call site can just write `IMPL->setText(String(...))` and then you don't need to name this "exported". The conversion from `NSString` does not **need** to be implicit.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -86,10 +87,39 @@ void HTMLScriptElement::didFinishInsertingNode()
+    if (value.hasException())
+        return value.releaseException();
+
+    Node::setTextContent(value.releaseReturnValue());
```
Another way to do this is to put `using HTMLElement::setTextContent` in the class definition so we overload rather than hiding the function inherited from `Node`. Then we can just call `setTextContent` here without the `Node::` prefix.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -86,10 +87,39 @@ void HTMLScriptElement::didFinishInsertingNode()
+    if (stringValueHolder.hasException())
+        return stringValueHolder.releaseException();
+
+    HTMLElement::setInnerText(stringValueHolder.releaseReturnValue());
```
Another way to do this is to put `using HTMLElement::setInnerText` in the class definition so we overload rather than hiding the function inherited from `HTMLElement`. Then we can just call `setInnerText` here without the `HTMLElement::` prefix.

lukewarlow commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -113,16 +143,26 @@ String HTMLScriptElement::crossOrigin() const
+String HTMLScriptElement::src() const
 {
-    return document().completeURL(sourceAttributeValue());
+    return getURLAttributeForBindings(WebCore::HTMLNames::srcAttr).string();
```
@darinadler One last question, the other changes were just related to your feedback.

I realised the API tests were failing due to the change from the Reflect, URL IDL to the existing src function. I've updated it to work how the binding code would have.

The existing src function only seemed to be used within the file itself inside of addSubresourceAttributeURLs so I've moved the original code there. Can you forsee any issues with this?

If it was used elsewhere I'd assume that the change from URL to String type would catch it in the compiler?

darinadler replied:
Yes, seems like you did exactly the right thing here. The only better thing would be if there was a way we could get the getter to be generated by the `[Reflect]` attribute, and still hand-write the setter. You might think we could do it with `[Reflect, CustomSetter]` but sadly a custom setter is too “custom” for what we want.

Moving the code inside `addSubresourceAttributeURLs` was exactly the right thing to do.

I agree that it’s unlikely `src()` is used anywhere else. I suggest you do a grep as well as counting on `URL` vs. `String`, since there are some types of super-simplistic calls that might still compile with the wrong type, so nice to be thorough.

lukewarlow replied:
I've had a look through and I can't seem to find a usage anywhere.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -86,10 +87,39 @@ void HTMLScriptElement::didFinishInsertingNode()
 void HTMLScriptElement::setText(String&& value)
 {
-    setTextContent(WTFMove(value));
+    setText(std::variant<RefPtr<TrustedScript>, String>(WTFMove(value)));
```
Shouldn’t this call `setTextContent` instead? We don’t need the trusted check here.

lukewarlow replied:
Ah yeah good point

darinadler commented on Source/WebCore/html/HTMLScriptElement.h:
```diff
@@ -37,8 +40,17 @@ class HTMLScriptElement final : public HTMLElement, public ScriptElement {
+    ExceptionOr<void> setTextContent(std::optional<std::variant<RefPtr<TrustedScript>, String>>&&);
+
+    using Node::setTextContent;
+    using HTMLElement::setInnerText;
```
It’s a bit strange to have this after our `setTextContent` function and before our `setInnerText` function. I suggest we put each one just above the function of the same name, in a shared paragraph:
```
using Node::setTextContent;
ExceptionOr<void> setTextContent(std::optional<std::variant<RefPtr<TrustedScript>, String>>&&);
```

lukewarlow committed:
Implement trusted types integration with script elements IDL
https://bugs.webkit.org/show_bug.cgi?id=269718

Reviewed by Ryosuke Niwa and Darin Adler.

Implement shadow properties for textContent and innerText on HTMLScriptElement.
Implement enforcement on src and text as well as the shadowed properties.

This doesn't implement any changes to the prepareScript function that will be addressed in a follow up.

Spec: https://w3c.github.io/trusted-types/dist/spec/#enforcement-in-scripts

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::trustedTypeCompliantString):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::setText):
(WebCore::HTMLScriptElement::setTextContent):
(WebCore::HTMLScriptElement::setInnerText):
(WebCore::HTMLScriptElement::src const):
(WebCore::HTMLScriptElement::setSrc):
(WebCore::HTMLScriptElement::addSubresourceAttributeURLs const):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.idl:
* Source/WebKitLegacy/mac/DOM/DOMHTMLScriptElement.mm:
(-[DOMHTMLScriptElement setText:]):

Canonical link: https://commits.webkit.org/278343@main
